syntax = "proto3";

service HashBrownie {
  rpc GetCode(GetCodeRequest) returns (Code) {}
  rpc GetAbi(GetAbiRequest) returns (Abi) {}
  rpc PutAbi(PutAbiRequest) returns (Empty) {}
  rpc GetLogs(GetLogsRequest) returns (Logs) {}
}

message Empty {}

message GetCodeRequest {
  string address = 1;
  optional int64 block = 2;
}

message GetAbiRequest {
  string address = 1;
}

message PutAbiRequest {
  string address = 1;
  Abi abi = 2;
}

message GetLogsRequest {
  message Topic {
    repeated string topics = 1;
  }
  repeated string addresses = 1;
  repeated Topic topics = 2;
  int64 from_block = 3;
  optional int64 to_block = 4;
}

message Code {
  bytes results = 1;
  string key = 2;
}

message Abi {
  message AbiEntry {
    message IO {
      string name = 1;
      string type = 2;
      string internalType = 3;
      optional bool indexed = 4;
      repeated IO components = 5;
    }
    string type = 1;
    optional string name = 2;
    optional bool anonymous = 3;
    optional int64 gas = 4;
    optional string stateMutability = 5;
    optional bool constant = 6;
    optional bool payable = 7;
    repeated IO inputs = 8;
    repeated IO outputs = 9;
  }

  repeated AbiEntry entries = 1;
  string contract_name = 2;
  string key = 3;
}

message Logs {
  message LogEntry {
    string address = 1;
    repeated bytes topics = 2;
    string data = 3;
    int64 blockNumber = 4;
    bytes transactionHash = 5;
    int64 transactionIndex = 6;
    bytes blockHash = 7;
    int64 logIndex = 8;
    bool removed = 9;
  }
  repeated LogEntry entries = 1;
  string key = 2;
}
